### DISCUSSION TITLE
Sungrow per Modbus

### DISCUSSION THREAD

**Discussion by holiday-sunrise** [solectrus/solectrus#2083] [Ideas]
Hallo Zusammen, ich habe das ganze bei einem Kollegen gesehen und bin begeistert. Ich habe eine Sungrow SH5.0RT.   

Diese hat eine Modbus Schnittstell kann ich das an Solectrus anbinden ?

Beste Grüße

Rainer

**Comment by ledermann** [comment]
Derzeitig unterstützt SOLECTRUS die Anbindung an SENEC (nativ), seit kurzem gibt es außerdem eine [MQTT-Schnittstelle](https://github.com/solectrus/mqtt-collector/). Eine direkte Modbus-Schnittstelle gibt es derzeitig nicht.

Insofern könnte der MQTT-Collector ein Ansatz sein, SOLECTRUS mit Deiner Sungrow-Anlage zu nutzen. Denkbar wäre, dass z.B ioBroker einen Adapter für Deine Anlage hat und die Messwerte dann über MQTT bereitstellt. Vielleicht gibt es aber auch ein anderes Tool, das eine Art Modbus-zu-MQTT-Wandlung ermöglicht.

**Comment by holiday-sunrise** [comment]
Hallo Georg,

 

vielen Dank für deinen Hinweis. Ich versuche es erstmal mit einem MQTT Adapter. Würde dann aber einen Modus Adapter schreiben.

 

Was mache ich gerade Falsch ?

 

Ich habe die Kennwörter alle geändert und die Compose um traefik erweitert

 




app_1                 | Waiting for InfluxDB on influxdb:8086 ...
influxdb_1            | ts=2023-06-18T12:33:00.226219Z lvl=info msg=Unauthorized log_id=0iVc_N1W000 error="authorization not found"
mqtt-collector_1      | Error while pushing to InfluxDB: unauthorized access
app_1                 | nc: bad address 'influxdb'
app_1                 | Waiting for InfluxDB on influxdb:8086 ...
influxdb_1            | ts=2023-06-18T12:33:05.231642Z lvl=info msg=Unauthorized log_id=0iVc_N1W000 error="authorization not found"
mqtt-collector_1      | Error while pushing to InfluxDB: unauthorized access
influxdb_1            | ts=2023-06-18T12:33:10.241511Z lvl=info msg=Unauthorized log_id=0iVc_N1W000 error="authorization not found"
mqtt-collector_1      | Error while pushing to InfluxDB: unauthorized access
app_1                 | nc: bad address 'influxdb'
app_1                 | Waiting for InfluxDB on influxdb:8086 ...
influxdb_1            | ts=2023-06-18T12:33:15.245778Z lvl=info msg=Unauthorized log_id=0iVc_N1W000 error="authorization not found"
mqtt-collector_1      | Error while pushing to InfluxDB: unauthorized access
app_1                 | nc: bad address 'influxdb'
app_1                 | Waiting for InfluxDB on influxdb:8086 ...
influxdb_1            | ts=2023-06-18T12:33:20.253880Z lvl=info msg=Unauthorized log_id=0iVc_N1W000 error="authorization not found"
mqtt-collector_1      | Error while pushing to InfluxDB: unauthorized access
app_1                 | nc: bad address 'influxdb'
app_1                 | Waiting for InfluxDB on influxdb:8086 ...
influxdb_1            | ts=2023-06-18T12:33:25.257738Z lvl=info msg=Unauthorized log_id=0iVc_N1W000 error="authorization not found"
mqtt-collector_1      | Error while pushing to InfluxDB: unauthorized access
^CGracefully stopping... (press Ctrl+C again to force)
Stopping solectrus_app_1                ... done
Stopping solectrus_mqtt-collector_1     ... done
Stopping solectrus_forecast-collector_1 ... done
Stopping solectrus_influxdb_1           ... done
Stopping solectrus_redis_1              ... done
Stopping solectrus_db_1                 ... done
 

 

Gesendet: Montag, 12. Juni 2023 um 14:46 Uhr
Von: "Georg Ledermann" ***@***.***>
An: "solectrus/solectrus" ***@***.***>
Cc: "Rainer Drexler" ***@***.***>, "Author" ***@***.***>
Betreff: Re: [solectrus/solectrus] Sungrow per Modbus (Discussion #2083)


 

Derzeitig unterstützt SOLECTRUS die Anbindung an SENEC (nativ), seit kurzem gibt es außerdem eine MQTT-Schnittstelle. Eine direkte Modbus-Schnittstelle gibt es derzeitig nicht.

Insofern könnte der MQTT-Collector ein Ansatz sein, SOLECTRUS mit Deiner Sungrow-Anlage zu nutzen. Denkbar wäre, dass z.B ioBroker einen Adapter für Deine Anlage hat und die Messwerte dann über MQTT bereitstellt. Vielleicht gibt es aber auch ein anderes Tool, das eine Art Modbus-zu-MQTT-Wandlung ermöglicht.

**Comment by ledermann** [comment]
Das Problem ist hier, dass InfluxDB nicht startet wegen "authorization not found". Dadurch laufen dann auch die App sowie der mqtt-collector nicht, weil beide ja auf InfluxDB zugreifen.

Ursache ist vermutlich die nachträgliche Änderung der Passwörter. Beim ersten Start von InfluxDB wird ein User mit dem angegebenen Kennwort erzeugt, das dann fix ist. 

Konkret: Folgende drei Werte in der `.env` dürfen **nach dem ersten Start von InfluxDB** nicht mehr geändert werden:

```.env
INFLUX_USERNAME=admin
INFLUX_PASSWORD=ExAmPl3PA55W0rD
INFLUX_ADMIN_TOKEN=my-super-secret-admin-token
```

Wenn Du die Passwörter dennoch ändern möchtest, muss die InfluxDB-Datenbank gelöscht werden, also z.B. so:

```bash
# Container herunterfahren
docker compose down

# Jetzt Influx-Credentials in .env bearbeiten (INFLUX_USERNAME, INFLUX_PASSWORD, INFLUX_ADMIN_TOKEN)
nano .env

# InfluxDB-Datenbank löschen (nur den Inhalt des Folders, nicht den Folder selbst)
rm -rf /home/pi/solectrus/influxdb/*

# Container neu starten, InfluxDB wird dabei neu angelegt
docker compose up
```

**Reply by holiday-sunrise** [reply]
Vielen Dank hat geholfen

**Comment by holiday-sunrise** [comment]
Zunächst mal. 

MQTT läuft fast wunderbar.

Netzverbrauch und Einspeisung wird nicht angezeigt. An welchem Wert hängt die Anzeige ?

Für welchen Teil der Anzeige werden die folgenden Mappings

MQTT_TOPIC_BAT_VOLTAGE=senec/0/ENERGY/GUI_BAT_DATA_VOLTAGE
MQTT_TOPIC_CASE_TEMP=senec/0/TEMPMEASURE/CASE_TEMP        

Battery Temp oder InverterTemp

MQTT_TOPIC_CURRENT_STATE=senec/0/ENERGY/STAT_STATE_Text
MQTT_TOPIC_MPP1_POWER=senec/0/PV1/MPP_POWER/0
MQTT_TOPIC_MPP2_POWER=senec/0/PV1/MPP_POWER/1
MQTT_TOPIC_MPP3_POWER=senec/0/PV1/MPP_POWER/2

**Reply by ledermann** [reply]
Netzverbrauch sowie Einspeisung werden über das gleiche Topic (`MQTT_TOPIC_GRID_POW`) abgewickelt und durch positive bzw. negative Werte dargestellt: Wenn der Wert negativ ist, dann ist es Einspeisung, wenn der Wert positiv ist, dann ist es Netzbezug.

Die Batteriespannung (`MQTT_TOPIC_BAT_VOLTAGE`) wird derzeitig in SOLECTRUS nicht angezeigt, die kann somit leer bleiben.

Die Temperatur (`MQTT_TOPIC_CASE_TEMP`) ist bei SENEC die Gehäusetemperatur. Ob Du lieber die Batterie-Temperatur oder die Wechselrichter-Temperatur (und sonst eine Temperatur) darstellen willst, kannst Du frei wählen.

Im Topic `MQTT_TOPIC_CURRENT_STATE` kann der aktuelle Status als Text (String) übergeben werden. Dieser wird dann rechts oben (anstelle von "LIVE") angezeigt.

Über die Topics MPP1/2/3 können die Erzeuger-Werte (in kW) der einzelnen Strings angegeben werden. Ob diese überhaupt vorliegen, hängt natürlich von der Anlage ab. Sie werden derzeitig aber noch nicht in SOLECTRUS angezeigt, das ist noch in Planung.

**Reply by holiday-sunrise** [reply]
Vielen lieben Dank.

**Reply by GalOR7** [reply]
Ja die MPPs zu sehen wäre klasse. Hier könnte man sehr schön sehen welcher String gerade wieviel Power macht 👍🏻

**Comment by holiday-sunrise** [comment]
Ich habe mal versucht einen Importer für Sungrow zu schreiben. https://github.com/holiday-sunrise/solectrus-sungrow-importer

als Beispiel habe ich den Senec importer genommen.


Problem beide funktionieren bei mir lokal im RubyMine nicht :-(  

Die Column per Name werden nicht gefunden.  Uhrzeit kann nicht geparsed werden. Auch nicht mit dem Senec Beispiel csv File aus deinem Projekt mit deinem Importer

Hast du eine Idee

**Reply by ledermann** [reply]
Wow, tolle Sache 👍 

Da einige Dinge zu ändern sind, ist es vermutlich einfacher, wenn ich das direkt selbst mache und das Repo in der SOLECTRUS-Org anlege. Letztendlich übernehme ich dann nur Dein CSV-File als Basis für die Tests. Ich hoffe, Du hast nichts dagegen.

Melde mich, wenn es einsatzfähig ist.

**Reply by holiday-sunrise** [reply]
Die Werte plus und minus habe ich noch nicht angepasst.War nur ein Erster test.  Also meine Test laufen nicht weil er die Zeit nicht parsen kann. Bin kein Ruby Entwickler. Noch nicht :-)

**Reply by holiday-sunrise** [reply]
Hätte es schon gern hinbekommen 😉

**Reply by ledermann** [reply]
Ich hatte zunächst Dein Repo ausprobiert und bin genau auf das gleiche Problem mit den Column Names gestoßen. Da ich die Ursache nicht gefunden hatte, habe ich es dann nochmal ganz neu auf Basis des `senec-importer` versucht. Das hat dann problemlos geklappt - inkl. der Plus/Minus-Sache für Grid und Battery und inklusive lauffähiger Tests und automatischer Erzeugung des Docker-Images.

Hier das Repo:
https://github.com/solectrus/sungrow-importer

Zu besseren Nachvollziehbarkeit meiner Änderungen liegt alles Relevante in diesem Commit:
https://github.com/solectrus/sungrow-importer/commit/1f4568348964e67e1bbb343edadc82d5cc0bbc1a

Bei der Verwendung bitte beachten, dass es eine `.env`-Variable `INFLUX_MEASUREMENT` gibt, damit man nicht den in diesem Kontext falschen Begriff "SENEC" verwenden muss.

Falls irgendwas nicht funktioniert, am besten ein Issue im Repo angelegen.

**Comment by holiday-sunrise** [comment]
ich habe mal alles von dir übernommen und es geht immer noch nicht. würde es gern verstehen.

```
Importing ./spec/data/TB_20230621.csv... F.

Failures:

  1) Import#run 
     Failure/Error: Time.parse("#{row[string]} CET").to_i
     
     ArgumentError:
       no time information in " CET"
     # ./app/solectrus_record.rb:72:in `parse_time'
     # ./app/solectrus_record.rb:18:in `time'
     # ./app/solectrus_record.rb:12:in `to_h'
     # ./app/import.rb:39:in `block in process'
     # ./app/import.rb:36:in `map'
     # ./app/import.rb:36:in `process'
     # ./app/import.rb:14:in `block in run'
     # ./app/import.rb:13:in `each'
     # ./app/import.rb:13:in `run'
     # ./spec/import_spec.rb:5:in `block (3 levels) in <top (required)>'
     # ./spec/import_spec.rb:7:in `block (3 levels) in <top (required)>'
```

bin unter Linux

**Reply by ledermann** [reply]
Verstehe ich leider auch nicht. Laufen denn die Tests **in meinem Repo** bei Dir? 

Mich irritiert, dass Du anfangs geschrieben hast, dass selbst der `senec-importer` lokal bei Dir nicht funktioniert. Hier bei mir (MacOS) läuft es und auf GitHub Actions (Linux) ebenfalls.
Was passiert, wenn Du das neue Docker-Image  (ghcr.io/solectrus/sungrow-importer) verwendest?

**Reply by holiday-sunrise** [reply]
Ja

**Reply by holiday-sunrise** [reply]
Der senec geht bei mir auch nicht

**Reply by ledermann** [reply]
Mit "geht nicht" meinst Du, dass die Tests lokal nicht fehlerfrei durchlaufen? Also folgendes klappt nicht?

```bash
git clone git@github.com:solectrus/senec-importer.git 
cd senec-importer
bundle install
bundle exec rake spec
```

Ich nehme an, dass die Verwendung des Docker-Images (gemäß README) aber auch bei Dir funktioniert.

**Reply by holiday-sunrise** [reply]
Ich hatte das lokale repro nochmal gelöscht dann ging es. Bei meinem nicht :-( Na ja ich suche weiter ich möchte es verstehen

Vielen Dank zum Import komme ich vielleicht morgen.

**Reply by holiday-sunrise** [reply]
Ich habe nun die cvs aus deinem Projekt in meins kopiert dann laufen die Tests durch. Ich verstehe es trotzdem nicht.

**Reply by ledermann** [reply]
Vielleicht irgendein Problem mit Zeilenumbrüchen (CR/LF) oder Encoding (UTF/ASCII, BOM)?

Ich bin übrigens dabei, aus den verschiedenen Import-Repos ein Generisches zu machen:
https://github.com/solectrus/senec-importer/pull/49

**Reply by ledermann** [reply]
Es lag tatsächlich am [BOM](https://de.wikipedia.org/wiki/Byte_Order_Mark), der in Deinem CSV-File enthalten war und Probleme verursachte. Hier mein Fix:
https://github.com/solectrus/senec-importer/pull/49/commits/0b0107f1880cc260f6d1118592b1891f67c70d36

**Reply by holiday-sunrise** [reply]
Super jetzt haben wir den Grund

**Comment by holiday-sunrise** [comment]
Ich bin noch dabei nun auf meinem Mac Ruby einzurichten. Leider noch ohne erfolg. brew install ... doch dann bekomme ich Fehler. Vielleicht liegt es an RubyMine und brew install ruby :-( 

Mein Projekt baut nun auch. So langsam komme ich klar mit dem Stack

**Reply by ledermann** [reply]
Auf dem Mac installiert man Ruby am besten über einen Versionsmanager wie z.B. [rbenv](https://formulae.brew.sh/formula/rbenv), also `brew install rbenv`, dann `rbenv install 3.2.2` für die momentan aktuelle Ruby-Version. Alternativen sind [rvm](https://formulae.brew.sh/formula/rvm) und [asdf](https://formulae.brew.sh/formula/asdf).

RubyMine habe ich zuletzt vor etwa 15 Jahren mal ausprobiert, dazu kann ich nichts sagen :-) 
Seit Längerem verwende ich [VS-Code](https://code.visualstudio.com/) für alle Sprachen.

**Comment by holiday-sunrise** [comment]
Hi Gorge, soll ich dev [senec-importer](https://github.com/solectrus/senec-importer) generic cvs testen ? Beste Grüße Rainer

**Reply by ledermann** [reply]
Gerne! Bin momentan mit anderen Projekten ausgelastet, daher stockt es hier ein wenig. Mein [Pull-Request](https://github.com/solectrus/senec-importer/pull/49) ist zwar fertig, aber noch nicht gemerged. Falls etwas nicht stimmen sollte, Kommentare bitte dort. 
Das Repo muss nach dem Merge auch umbenannt werden in `csv-importer`.
